README.md
# Agegate

A simple function that verifies a date of birth against a country's legal drinking age.

```
npm i agegate
```

## Usage

```js
import agegate from "agegate";

var user = {
  dateOfBirth: new Date("2015-02-14"), // strings are also accepted
  country: "US",
};

var isLegal = agegate(user.dateOfBirth, user.country); // false
```

:warning: If an invalid date is supplied, the result will be falsy. If an invalid country code is supplied, it will validate against a default legal drinking age of 18.

### Use with frameworks (e.g. React)

In order to use this library with frontend UI frameworks, the underlying dataset used to validate is also exported.

```js
import React, { useState } from "react";
import agegate, { getData } from "agegate";

const countries = getData();

function Modal() {
  const [date, setDate] = useState("");
  const [country, setCountry] = useState(countries[0].code);
  const [legal, setLegal] = useState(false);

  const submitHandler = e => {
    e.preventDefault();

    if (date && country) {
      const result = agegate(new Date(date), country);
      setLegal(result);
    }
  };

  return (
    <div>
      <form onSubmit={submitHandler}>
        <h3>Enter your date of birth</h3>
        <input
          type="date"
          value={date}
          onChange={e => setDate(e.target.value)}
        />

        <h3>Enter your country</h3>
        <select value={country} onChange={e => setCountry(e.target.value)}>
          {countries.map(({ code, name }) => (
            <option key={name} value={code}>
              {name}
            </option>
          ))}
        </select>

        <button type="submit">Submit</button>
      </form>

      <p style={{ color: legal ? "green" : "red" }}>
        RESULT: You are {legal ? "" : "NOT"} old enough!
      </p>
    </div>
  );
}
```

* * *

Please file a [new issue](https://github.com/izolate/agegate/issues/new) if you find any inconsistencies in the countries dataset.


CHANGELOG.md
# 4.0.0 (2020-09-11)

* Completely refactored library into a single validator function. Removed any elements such as cookies, HTML forms, etc. This new narrow focus modernizes the library and allows it to be used with frontend UI frameworks too.

# 3.0.0 (2016-04-15)
* Changes key name of cookie-related options

# 2.1.0 (2016-04-15)
* Adds option to set cookie name
* Code style improvements

# 1.0.0 (2015-04-27)
### Features
* Added the rest of the countries to the data file

### Bugs
* Re-written the method to serialize form data to be more secure
* Corrected the method for retrieval of input's checked value
* Renamed property `defaults` to `options`
* Indicated the base number on `parseInt()`

# 0.3.0 (2015-04-23)
### Features
* Add `data` option to allow override of default country drinking age data
* JSDoc comments
* Test for data validation

### Bugs
* Default `month` and `day` to `1`, if not specified
* More specific variable names to avoid duplication
* `parseInt()` called on numeric form data

# 0.2.0 (2015-04-20)
### Features
* Build UMD modules, instead of purely CommonJS
* Test for saving cookies
* Styled example page
* `.npmignore` and `.jshintrc`

### Bugs
* `verify(data)` method responds `boolean`

# 0.1.0
Initial commit


test/test.js
var assert = chai.assert;

describe('AgeGate', function() {
  var form = document.querySelector('form');

  describe('Instantiate class', function() {
    var gate = new AgeGate({
      form: form,
      age: 21,
      cookieExpiry: Infinity,
      cookieName: 'custom_cookie_name',
      cookieDomain: 'custom_cookie_domain',
      cookiePath: 'custom_cookie_path',
      countries: true,
    }, null);

    it('should save options', function() {
      assert.equal(gate.options.form, form);
      assert.strictEqual(gate.options.cookieExpiry, Infinity);
      assert.strictEqual(gate.options.cookieName, 'custom_cookie_name');
      assert.strictEqual(gate.options.cookieDomain, 'custom_cookie_domain');
      assert.strictEqual(gate.options.cookiePath, 'custom_cookie_path');
      assert.strictEqual(gate.options.countries, true);
      assert.strictEqual(gate.options.age, 21);
    });

    it('should override default age with supplied age', function() {
      assert.isAbove(gate.legalAge, 18);
    });

    it('should populate the <select> with countries', function() {
      var select = form.querySelector('select');
      assert.isAbove(select.children.length, 0);
    });
  });

  describe('Validation', function() {
    it('should validate supplied data', function() {
      var gate = new AgeGate({
        age: 18,
        form: form,
        cookieName: 'custom_cookie_name'
      }, function(err) {
        if (err)
          assert.equal(err.message, '[AgeGate] Supplied data is invalid');
      });

      var valid = [{name: 'United Kingdom', code: 'UK', age: 18}];
      var invalid = [{this: 1 , is: 2, invalid: 3}];

      assert.equal(gate.validateData(valid), valid);
      gate.validateData(invalid);
    });

    it('should validate ages correctly', function() {
      var gate = new AgeGate({
        age: 18,
        form: form,
        cookieName: 'custom_cookie_name'
      }, function() {});

      var today = new Date();
      var old = {
        year: today.getFullYear()-20, month: today.getMonth()+1, day: today.getDate()
      };
      var young = { year: old.year+10, month: old.month, day: old.day };
      assert.isTrue(gate.verify(old), 'Old enough');
      assert.isFalse(gate.verify(young), 'Too young');
    });
  });

  describe('Cookies', function() {
    it('should save cookies', function() {
      assert.isTrue(Cookies.hasItem('custom_cookie_name'));
    });
  });
});



test/agegate.js
window.Cookies = require('../dist/cookies.js');
window.AgeGate = require('../dist/index.js');


src/index.test.js
import agegate, { getData } from "./index.js";

import test from "ava";

test("returns true if age is above legal age", async (t) => {
  const now = new Date();
  now.setFullYear(now.getFullYear() - 22);
  const result = agegate(now, "US");
  t.truthy(result);
});

test("returns false if age is below legal age", async (t) => {
  const now = new Date();
  now.setFullYear(now.getFullYear() - 19);
  const result = agegate(now, "US");
  t.falsy(result);
});

test("converts strings to Date", async (t) => {
  const value = "1969-05-20";
  const result = agegate(value, "US");
  t.truthy(result);
});

test("returns falsy if date is invalid", async (t) => {
  const value = "INVALID_DATE";
  const result = agegate(value, "US");
  t.falsy(result);
});

test("uses 18 as default drinking age", async (t) => {
  const now = new Date();
  now.setFullYear(now.getFullYear() - 10);
  const result = agegate(now, "INVALID_COUNTRY_CODE");
  t.falsy(result);
});

test("accepts lowercase country codes", async (t) => {
  const result = agegate("1970-01-01", "us");
  t.truthy(result);
});


test("country data is exported", async (t) => {
  t.is(getData().length, 177);
});


src/data.js
export var Africa = [
  { code: 'DZ', name: 'Algeria', age: 18 },
  { code: 'AO', name: 'Angola', age: 18 },
  { code: 'DZ', name: 'Botswana', age: 18 },
  { code: 'BI', name: 'Burundi', age: 18 },
  { code: 'CM', name: 'Cameroon', age: 21 },
  { code: 'CV', name: 'Cape Verde', age: 18 },
  { code: 'CF', name: 'Central African Republic', age: 18 },
  { code: 'KM', name: 'Comoros', age: 0 },
  { code: 'CD', name: 'Democratic Republic of the Congo', age: 18 },
  { code: 'EG', name: 'Egypt', age: 21 },
  { code: 'GQ', name: 'Equatorial Guinea', age: 0 },
  { code: 'ER', name: 'Eritrea', age: 18 },
  { code: 'ET', name: 'Ethiopia', age: 18 },
  { code: 'GA', name: 'Gabon', age: 18 },
  { code: 'GM', name: 'Gambia', age: 18 },
  { code: 'GH', name: 'Ghana', age: 18 },
  { code: 'GW', name: 'Guinea-Bissau', age: 0 },
  { code: 'KE', name: 'Kenya', age: 18 },
  { code: 'LS', name: 'Lesotho', age: 18 },
  { code: 'LY', name: 'Libya', age: Infinity },
  { code: 'MW', name: 'Malawi', age: 18 },
  { code: 'MU', name: 'Mauritius', age: 18 },
  { code: 'MA', name: 'Morocco', age: 16 },
  { code: 'MZ', name: 'Mozambique', age: 18 },
  { code: 'NA', name: 'Namibia', age: 18 },
  { code: 'NE', name: 'Niger', age: 18 },
  { code: 'NG', name: 'Nigeria', age: 18 },
  { code: 'CG', name: 'Republic of the Congo', age: 18 },
  { code: 'RW', name: 'Rwanda', age: 18 },
  { code: 'SN', name: 'Senegal', age: 0 },
  { code: 'SC', name: 'Seychelles', age: 18 },
  { code: 'SO', name: 'Somalia', age: Infinity },
  { code: 'ZA', name: 'South Africa', age: 18 },
  { code: 'SS', name: 'South Sudan', age: 18 },
  { code: 'SD', name: 'Sudan', age: Infinity },
  { code: 'SZ', name: 'Swaziland', age: 18 },
  { code: 'TZ', name: 'Tanzania', age: 18 },
  { code: 'TG', name: 'Togo', age: 18 },
  { code: 'TN', name: 'Tunisia', age: 18 },
  { code: 'UG', name: 'Uganda', age: 18 },
  { code: 'ZM', name: 'Zambia', age: 18 },
  { code: 'ZW', name: 'Zimbabwe', age: 18 }
]

export var America = [
  { code: 'AG', name: 'Antigua and Barbuda', age: 16 },
  { code: 'AR', name: 'Argentina', age: 18 },
  { code: 'BS', name: 'Bahamas', age: 18 },
  { code: 'BZ', name: 'Belize', age: 18 },
  { code: 'BM', name: 'Bermuda', age: 18 },
  { code: 'BO', name: 'Bolivia', age: 18 },
  { code: 'BR', name: 'Brazil', age: 18 },
  { code: 'VG', name: 'British Virgin Islands', age: 16 },
  { code: 'CA', name: 'Canada', age: 19 },
  { code: 'KY', name: 'Cayman Islands', age: 18 },
  { code: 'CL', name: 'Chile', age: 18 },
  { code: 'CO', name: 'Colombia', age: 18 },
  { code: 'CR', name: 'Costa Rica', age: 18 },
  { code: 'CU', name: 'Cuba', age: 18 },
  { code: 'DO', name: 'Dominican Republic', age: 18 },
  { code: 'EC', name: 'Ecuador', age: 18 },
  { code: 'SV', name: 'El Salvador', age: 18 },
  { code: 'FK', name: 'Falkland Islands', age: 18 },
  { code: 'GT', name: 'Guatemala', age: 18 },
  { code: 'GY', name: 'Guyana', age: 18 },
  { code: 'HT', name: 'Haiti', age: 16 },
  { code: 'HN', name: 'Honduras', age: 18 },
  { code: 'JM', name: 'Jamaica', age: 16 },
  { code: 'MX', name: 'Mexico', age: 18 },
  { code: 'NI', name: 'Nicaragua', age: 18 },
  { code: 'PA', name: 'Panama', age: 18 },
  { code: 'PY', name: 'Paraguay', age: 20 },
  { code: 'PE', name: 'Peru', age: 18 },
  { code: 'PR', name: 'Puerto Rico', age: 18 },
  { code: 'TT', name: 'Trinidad and Tobago', age: 18 },
  { code: 'US', name: 'United States', age: 21 },
  { code: 'VI', name: 'United States Virgin Islands', age: 18 },
  { code: 'UY', name: 'Uruguay', age: 18 },
  { code: 'VE', name: 'Venezuela', age: 18 }
]

export var Asia = [
  { code: 'AF', name: 'Afghanistan', age: Infinity },
  { code: 'BD', name: 'Bangladesh', age: Infinity },
  { code: 'BN', name: 'Brunei', age: Infinity },
  { code: 'KH', name: 'Cambodia', age: 0 },
  { code: 'CN', name: 'China', age: 18 },
  { code: 'HK', name: 'Hong Kong', age: 18 },
  { code: 'IN', name: 'India', age: 18 },
  { code: 'ID', name: 'Indonesia', age: 21 },
  { code: 'IR', name: 'Iran', age: 18 },
  { code: 'JQ', name: 'Iraq', age: 18 },
  { code: 'IL', name: 'Israel', age: 18 },
  { code: 'JP', name: 'Japan', age: 20 },
  { code: 'JO', name: 'Jordan', age: 18 },
  { code: 'KZ', name: 'Kazakhstan', age: 21 },
  { code: 'KW', name: 'Kuwait', age: Infinity },
  { code: 'KG', name: 'Kyrgyzstan', age: 18 },
  { code: 'LB', name: 'Lebanon', age: 18 },
  { code: 'MO', name: 'Macau', age: 18 },
  { code: 'MY', name: 'Malaysia', age: 18 },
  { code: 'MV', name: 'Maldives', age: 18 },
  { code: 'MN', name: 'Mongolia', age: 18 },
  { code: 'NP', name: 'Nepal', age: 18 },
  { code: 'KP', name: 'North Korea', age: 18 },
  { code: 'OM', name: 'Oman', age: 21 },
  { code: 'PK', name: 'Pakistan', age: 21 },
  { code: 'PS', name: 'Palestine', age: 16 },
  { code: 'PH', name: 'Philippines', age: 18 },
  { code: 'QA', name: 'Qatar', age: 21 },
  { code: 'SA', name: 'Saudi Arabia', age: Infinity },
  { code: 'SG', name: 'Singapore', age: 18 },
  { code: 'KR', name: 'South Korea', age: 19 },
  { code: 'LK', name: 'Sri Lanka', age: 21 },
  { code: 'SY', name: 'Syria', age: 18 },
  { code: 'TW', name: 'Taiwan', age: 18 },
  { code: 'TJ', name: 'Tajikistan', age: 21 },
  { code: 'TH', name: 'Thailand', age: 21 },
  { code: 'TM', name: 'Turkmenistan', age: 18 },
  { code: 'AE', name: 'United Arab Emirates', age: 21 },
  { code: 'VN', name: 'Vietnam', age: 0 },
  { code: 'YE', name: 'Yemen', age: Infinity }
]

export var Europe = [
  { code: 'AL', name: 'Albania', age: 18 },
  { code: 'AM', name: 'Armenia', age: 18 },
  { code: 'AT', name: 'Austria', age: 18 },
  { code: 'AZ', name: 'Azerbaijan', age: 18 },
  { code: 'BY', name: 'Belarus', age: 18 },
  { code: 'BE', name: 'Belgium', age: 18 },
  { code: 'BA', name: 'Bosnia and Herzegovina', age: 18 },
  { code: 'BG', name: 'Bulgaria', age: 18 },
  { code: 'HR', name: 'Croatia', age: 18 },
  { code: 'CY', name: 'Cyprus', age: 17 },
  { code: 'CZ', name: 'Czech Republic', age: 18 },
  { code: 'DK', name: 'Denmark', age: 18 },
  { code: 'EE', name: 'Estonia', age: 18 },
  { code: 'FI', name: 'Finland', age: 18 },
  { code: 'FR', name: 'France', age: 18 },
  { code: 'GE', name: 'Georgia', age: 16 },
  { code: 'DE', name: 'Germany', age: 18 },
  { code: 'GI', name: 'Gibraltar', age: 18 },
  { code: 'GR', name: 'Greece', age: 18 },
  { code: 'HU', name: 'Hungary', age: 18 },
  { code: 'IS', name: 'Iceland', age: 20 },
  { code: 'IE', name: 'Ireland', age: 18 },
  { code: 'IT', name: 'Italy', age: 18 },
  { code: 'XK', name: 'Kosovo', age: 18 },
  { code: 'LV', name: 'Latvia', age: 18 },
  { code: 'LI', name: 'Liechtenstein', age: 18 },
  { code: 'LT', name: 'Lithuania', age: 18 },
  { code: 'LU', name: 'Luxembourg', age: 16 },
  { code: 'MK', name: 'Macedonia', age: 18 },
  { code: 'MT', name: 'Malta', age: 17 },
  { code: 'MD', name: 'Moldova', age: 16 },
  { code: 'ME', name: 'Montenegro', age: 18 },
  { code: 'NL', name: 'Netherlands', age: 18 },
  { code: 'NO', name: 'Norway', age: 20 },
  { code: 'PL', name: 'Poland', age: 18 },
  { code: 'PT', name: 'Portugal', age: 18 },
  { code: 'RO', name: 'Romania', age: 18 },
  { code: 'RU', name: 'Russia', age: 18 },
  { code: 'RS', name: 'Serbia', age: 18 },
  { code: 'SK', name: 'Slovakia', age: 18 },
  { code: 'SI', name: 'Slovenia', age: 18 },
  { code: 'ES', name: 'Spain', age: 18 },
  { code: 'SE', name: 'Sweden', age: 18 },
  { code: 'CH', name: 'Switzerland', age: 18 },
  { code: 'TR', name: 'Turkey', age: 18 },
  { code: 'UA', name: 'Ukraine', age: 18 },
  { code: 'GB', name: 'United Kingdom', age: 18 }
]

export var Oceania = [
  { code: 'AS', name: 'American Samoa', age: 21 },
  { code: 'AU', name: 'Australia', age: 18 },
  { code: 'FJ', name: 'Fiji', age: 18 },
  { code: 'GU', name: 'Guam', age: 21 },
  { code: 'FM', name: 'Federated States of Micronesia', age: 21 },
  { code: 'NZ', name: 'New Zealand', age: 18 },
  { code: 'MP', name: 'Northern Mariana Islands', age: 21 },
  { code: 'PW', name: 'Palau', age: 21 },
  { code: 'PG', name: 'Papua New Guinea', age: 18 },
  { code: 'WS', name: 'Samoa', age: 18 },
  { code: 'SB', name: 'Solomon Islands', age: 21 },
  { code: 'TK', name: 'Tokelau', age: 18 },
  { code: 'TO', name: 'Tonga', age: 21 },
  { code: 'VU', name: 'Vanuatu', age: 18 }
]


src/index.js
import data from "./countries.js";

// Flatten continent-grouped country data into a simpler map.
export const getData = () => Object.values(data).reduce((a, b) => a.concat(b));
const countries = getData();

const DEFAULT_LEGAL_AGE = 18;

// Age calculator by Lucas Janon.
// https://stackoverflow.com/a/50827764/362136
const getAge = (date) =>
  Math.floor((new Date() - new Date(date).getTime()) / 3.15576e10);

function agegate(date, countryCode) {
  const age = getAge(date);
  const country = countries.find((c) => c.code === countryCode.toUpperCase());

  // null country age value indicates alcohol prohibition
  if ((country && country.age === null) || isNaN(age)) {
    return false;
  }

  const ageLimit = country ? country.age : DEFAULT_LEGAL_AGE;
  return age >= ageLimit;
}

export default agegate;


src/cookies.js
/*\
|*|
|*|  :: cookies.js ::
|*|
|*|  A complete cookies reader/writer framework with full unicode support.
|*|
|*|  Revision #1 - September 4, 2014
|*|
|*|  https://developer.mozilla.org/en-US/docs/Web/API/document.cookie
|*|  https://developer.mozilla.org/User:fusionchess
|*|
|*|  This framework is released under the GNU Public License, version 3 or later.
|*|  http://www.gnu.org/licenses/gpl-3.0-standalone.html
|*|
|*|  Syntaxes:
|*|
|*|  * cookies.setItem(name, value[, end[, path[, domain[, secure]]]])
|*|  * cookies.getItem(name)
|*|  * cookies.removeItem(name[, path[, domain]])
|*|  * cookies.hasItem(name)
|*|  * cookies.keys()
|*|
\*/

export default {
  getItem(sKey) {
    if (!sKey) { return null; }
    return decodeURIComponent(document.cookie.replace(new RegExp('(?:(?:^|.*;)\\s*' + encodeURIComponent(sKey).replace(/[\-\.\+\*]/g, '\\$&') + '\\s*\\=\\s*([^;]*).*$)|^.*$'), '$1')) || null;
  },

  setItem(sKey, sValue, vEnd, sPath, sDomain, bSecure) {
    if (!sKey || /^(?:expires|max\-age|path|domain|secure)$/i.test(sKey)) { return false; }
    var sExpires = '';
    if (vEnd) {
      switch (vEnd.constructor) {
        case Number:
          sExpires = vEnd === Infinity ? '; expires=Fri, 31 Dec 9999 23:59:59 GMT' : '; max-age=' + vEnd;
          break;
        case String:
          sExpires = '; expires=' + vEnd;
          break;
        case Date:
          sExpires = '; expires=' + vEnd.toUTCString();
          break;
      }
    }
    document.cookie = encodeURIComponent(sKey) + '=' + encodeURIComponent(sValue) + sExpires + (sDomain ? '; domain=' + sDomain : '') + (sPath ? '; path=' + sPath : '') + (bSecure ? '; secure' : '');
    return true;
  },

  removeItem(sKey, sPath, sDomain) {
    if (!this.hasItem(sKey)) { return false; }
    document.cookie = encodeURIComponent(sKey) + '=; expires=Thu, 01 Jan 1970 00:00:00 GMT' + (sDomain ? '; domain=' + sDomain : '') + (sPath ? '; path=' + sPath : '');
    return true;
  },

  hasItem(sKey) {
    if (!sKey) { return false; }
    return (new RegExp('(?:^|;\\s*)' + encodeURIComponent(sKey).replace(/[\-\.\+\*]/g, '\\$&') + '\\s*\\=')).test(document.cookie);
  },

  keys() {
    var aKeys = document.cookie.replace(/((?:^|\s*;)[^\=]+)(?=;|$)|^\s*|\s*(?:\=[^;]*)?(?:\1|$)/g, '').split(/\s*(?:\=[^;]*)?;\s*/);
    for (var nLen = aKeys.length, nIdx = 0; nIdx < nLen; nIdx++) { aKeys[nIdx] = decodeURIComponent(aKeys[nIdx]); }
    return aKeys;
  }
};


example/agegate-compiled.js
(function e(t,n,r){function s(o,u){if(!n[o]){if(!t[o]){var a=typeof require=="function"&&require;if(!u&&a)return a(o,!0);if(i)return i(o,!0);var f=new Error("Cannot find module '"+o+"'");throw f.code="MODULE_NOT_FOUND",f}var l=n[o]={exports:{}};t[o][0].call(l.exports,function(e){var n=t[o][1][e];return s(n?n:e)},l,l.exports,e,t,n,r)}return n[o].exports}var i=typeof require=="function"&&require;for(var o=0;o<r.length;o++)s(r[o]);return s})({1:[function(require,module,exports){
'use strict';

Object.defineProperty(exports, '__esModule', {
  value: true
});
/*
 * Africa
 */
var africa = [{
  code: 'DZ',
  name: 'Algeria',
  age: 18
}, {
  code: 'AO',
  name: 'Angola',
  age: 18
}, {
  code: 'DZ',
  name: 'Botswana',
  age: 18
}, {
  code: 'BI',
  name: 'Burundi',
  age: 18
}, {
  code: 'CM',
  name: 'Cameroon',
  age: 21
}, {
  code: 'CV',
  name: 'Cape Verde',
  age: 18
}, {
  code: 'CF',
  name: 'Central African Republic',
  age: 18
}, {
  code: 'KM',
  name: 'Comoros',
  age: 0
}, {
  code: 'EG',
  name: 'Egypt',
  age: 21
}, {
  code: 'GQ',
  name: 'Equatorial Guinea',
  age: 0
}, {
  code: 'ER',
  name: 'Eritrea',
  age: 18
}, {
  code: 'ET',
  name: 'Ethiopia',
  age: 18
}, {
  code: 'GA',
  name: 'Gabon',
  age: 18
}, {
  code: 'GM',
  name: 'Gambia',
  age: 18
}, {
  code: 'GH',
  name: 'Ghana',
  age: 18
}, {
  code: 'GW',
  name: 'Guinea-Bissau',
  age: 0
}, {
  code: 'KE',
  name: 'Kenya',
  age: 18
}];
/*
TODO Get drunk and finish the rest manually

Lesotho 18
Libya Infinity
Malawi 18
Mauritius 18
Morocco 16
Mozambique 18
Namibia 18
Niger 18
Nigeria 18
Republic of the Congo 18
Rwanda 18
Senegal 0
Seychelles 18
Somalia Illegal
South Africa 18
South Sudan 18
Sudan Infinity
Swaziland 18
Tanzania 18
Togo 18
Tunisia 18
Uganda 18
Zambia 18
Zimbabwe 18
*/

exports['default'] = {
  Africa: africa
  // TODO the rest of the continents
};
module.exports = exports['default'];

},{}],2:[function(require,module,exports){
'use strict';

var _interopRequireWildcard = function _interopRequireWildcard(obj) {
  return obj && obj.__esModule ? obj : { 'default': obj };
};

var _classCallCheck = function _classCallCheck(instance, Constructor) {
  if (!(instance instanceof Constructor)) {
    throw new TypeError('Cannot call a class as a function');
  }
};

var _createClass = (function () {
  function defineProperties(target, props) {
    for (var i = 0; i < props.length; i++) {
      var descriptor = props[i];descriptor.enumerable = descriptor.enumerable || false;descriptor.configurable = true;if ('value' in descriptor) descriptor.writable = true;Object.defineProperty(target, descriptor.key, descriptor);
    }
  }return function (Constructor, protoProps, staticProps) {
    if (protoProps) defineProperties(Constructor.prototype, protoProps);if (staticProps) defineProperties(Constructor, staticProps);return Constructor;
  };
})();

Object.defineProperty(exports, '__esModule', {
  value: true
});

var _data = require('./data');

var _data2 = _interopRequireWildcard(_data);

// Default configuration
var defaults = {
  html: ['input', 'p', 'select'],
  year: function year() {
    // calculate 18 years (216 months) prior to today
    var d = new Date();
    d.setMonth(d.getMonth() - 216);
    return d.getFullYear();
  }
};

var AgeGate = (function () {
  function AgeGate(opt) {
    var _this = this;

    _classCallCheck(this, AgeGate);

    // set defaults
    this.data = _data2['default'];
    this.defaults = {
      year: opt.startingDate || defaults.year()
    };

    if (typeof opt.form === 'undefined') throw new ReferenceError('No form HTML element defined');else this.form = opt.form;

    // ensure form contains required HTML elements
    defaults.html.forEach(function (elem) {
      try {
        _this.form.querySelector(elem);
      } catch (err) {
        throw new ReferenceError('<form> doesn\'t contain <' + elem + '> element');
      }
    });
  }

  _createClass(AgeGate, [{
    key: 'render',
    value: function render() {
      console.log('showing ageGate');
      this.populateData();
    }
  }, {
    key: 'populateData',
    value: function populateData() {
      var _this2 = this;

      // select
      Object.keys(_data2['default']).forEach(function (continent) {
        var group = document.createElement('optgroup');
        group.label = continent;

        for (var i = 0; i < _data2['default'][continent].length; i++) {
          var option = document.createElement('option'),
              country = _data2['default'][continent][i];
          option.value = country.code;
          option.textContent = country.name;
          option.dataset.code = country.code;
          option.dataset.name = country.name;
          option.dataset.age = country.age;
          group.appendChild(option);
        }

        _this2.form.querySelector('select').appendChild(group);
      });

      // input
      console.log(this.form.querySelector('input'));
      this.form.querySelector('input').value = this.defaults.year;
    }
  }]);

  return AgeGate;
})();

exports['default'] = AgeGate;
module.exports = exports['default'];

},{"./data":1}],3:[function(require,module,exports){
'use strict';

var _interopRequireWildcard = function (obj) { return obj && obj.__esModule ? obj : { 'default': obj }; };

var _AgeGate = require('../dist/index');

var _AgeGate2 = _interopRequireWildcard(_AgeGate);

var gate = window.gate = new _AgeGate2['default']({
  form: document.querySelector('form.agegate'),
  callback: function callback(err, res) {
    console.log(err, res);
  }
});

document.addEventListener('DOMContentLoaded', function () {
  gate.render();
});

},{"../dist/index":2}]},{},[3]);
